SingularityService/src/main/java/com/hubspot/singularity/data/ApiCache.java

```diff
@@ -0,0 +1,94 @@
+package com.hubspot.singularity.data;
+
+import com.google.inject.Inject;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ApiCache<K, V> {
+  private static final Logger LOG = LoggerFactory.getLogger(ApiCache.class);
+
+  public final boolean isEnabled;
+  private final AtomicReference<Map<K, V>> zkValues;
+  private final Supplier<Map<K, V>> supplyMap;
+  private final int cacheTtl;
+  private final ScheduledExecutorService executor;
+
+  private ScheduledFuture<?> reloadingFuture;
+
+  @Inject
+  public ApiCache(
+    boolean isEnabled,
+    int cacheTtl,
+    Supplier<Map<K, V>> supplyMap,
+    ScheduledExecutorService executor
+  ) {
+    this.isEnabled = isEnabled;
+    this.supplyMap = supplyMap;
+    this.zkValues = new AtomicReference<>(new HashMap<>());
+    this.cacheTtl = cacheTtl;
+    this.executor = executor;
+  }
+
+  public void startReloader() {
```

Comment:
should this also synchronously perform the first fetch I wonder? Would avoid the case where getAll was incorrectly empty or get(K key) was null